python/ray/_private/accelerator.py

```diff
@@ -0,0 +1,111 @@
+    function will check if the number of NeuronCore is greater than the
+    number of visible NeuronCore and raise an error if it is true.
+
+    If the number of NeuronCore is greater than the number of visible
+    NeuronCore, this function will raise an error.
+
+    Lastly, update accelerator_type and num_neuron_cores in resources.
+
+    Args:
+        resources: Resources dictionary to be updated with
+        NeuronCore accelerator type and custom resources(num_neuron_cores).
+
+    Raises:
+        ValueError: If the number of NeuronCore is greater than the number of
+            visible NeuronCore.
+    """
+    import ray._private.ray_constants as ray_constants
+    import ray._private.utils as utils
+
+    # AWS NeuronCore detection and configuration
+    # 1. Check if the user specified num_neuron_cores in resources
+    num_neuron_cores = resources.get(ray_constants.NUM_NEURON_CORES, None)
+    # 2. Check if the user specified NEURON_RT_VISIBLE_CORES
+    neuron_core_ids = utils.get_aws_neuron_core_visible_ids()
+    if (
+        num_neuron_cores is not None
+        and neuron_core_ids is not None
+        and num_neuron_cores > len(neuron_core_ids)
+    ):
+        raise ValueError(
+            f"Attempting to start raylet with {num_neuron_cores} "
+            f"neuron cores, but NEURON_RT_VISIBLE_CORES contains "
+            f"{neuron_core_ids}."
+        )
+    # 3. Auto-detect num_neuron_cores if not specified in resources
+    if num_neuron_cores is None:
+        num_neuron_cores = _autodetect_aws_neuron_cores()
+        # Don't use more neuron cores than allowed by NEURON_RT_VISIBLE_CORES.
+        if num_neuron_cores is not None and neuron_core_ids is not None:
+            num_neuron_cores = min(num_neuron_cores, len(neuron_core_ids))
+    if num_neuron_cores is not None:
+        # 4. Update accelerator_type and num_neuron_cores with
+        # number of neuron cores detected or configured.
+        resources.update(
+            {
+                ray_constants.NUM_NEURON_CORES: num_neuron_cores,
+                utils.get_neuron_core_constraint_name(): num_neuron_cores,
+            }
+        )
+
+
+def _autodetect_aws_neuron_cores() -> Optional[int]:
+    """
+    Attempt to detect the number of Neuron cores on this machine.
+
+    Returns:
+        The number of Neuron cores if any were detected, otherwise None.
+    """
+    result = None
+    if sys.platform.startswith("linux") and os.path.isdir("/opt/aws/neuron/bin/"):
+        result = _get_neuron_core_count()
+    return result
+
+
+def _get_neuron_core_count() -> int:
+    """Get the number of Neuron cores on a machine based on neuron_path.
+
+    Returns:
+        The number of Neuron cores on this machine (Default to 0).
+    """
+    neuron_path = "/opt/aws/neuron/bin/"
+    nc_count: int = 0
+    result = subprocess.run(
+        [os.path.join(neuron_path, "neuron-ls"), "--json-output"],
```

Comment:
What is the versioning contract here? With CUDA we rely on a Python package which is versioned and we can install with a Ray installation. Seems this will couple the Ray version with neuron-ls version.

(Sorry if this was discussed already)

Reply:
Good Q, today this tool comes as a [pre-requisite installation](https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/announcements/neuron2.x/neuron230-packages-changes.html) for NeuronCore support (on EC2 machine). Unfortunately, it's not available as pip package but available as rpm/deb (will take as a feature-request). IMO, the coupling is specific to resource environment (EC2 instance).

We also have an [open-item](https://github.com/ray-project/ray/issues/38473) (work-in-progress) to expose this information using public EC2 API (but this would put the burden on customer to include IAM permissions for EC2 role to call the API - but reliable).